src/QMCWaveFunctions/Jastrow/J1Spin.h

```diff
@@ -0,0 +1,854 @@
+  ///reference to the sources (ions)
+  const ParticleSet& Ions;
+  ///reference to the target (elecs)
+  const ParticleSet& Elecs;
+
+  ///number of variables this object handles
+  int NumVars;
+  ///variables handled by this orbital
+  opt_variables_type myVars;
+
+  valT curAt;
+  valT curLap;
+  posT curGrad;
+
+  ///\f$Vat[i] = sum_(j) u_{i,j}\f$
+  Vector<valT> Vat;
+  aligned_vector<valT> U, dU, d2U, d3U;
+  aligned_vector<valT> DistCompressed;
+  aligned_vector<int> DistIndice;
+  Vector<posT> Grad;
+  Vector<valT> Lap;
+  ///Container for \f$F[ig*NIons+jg]\f$
+  std::vector<FT*> J1Functors;
+  ///container for the unique Jastrow functions
+  std::map<std::string, std::unique_ptr<FT>> J1UniqueFunctors;
+
+  std::vector<std::pair<int, int>> OffSet;
+  Vector<RealType> dLogPsi;
+  typedef ParticleAttrib<QTFull::GradType> WavefunctionFirstDerivativeType;
+  typedef ParticleAttrib<QTFull::ValueType> WavefunctionSecondDerivativeType;
+  std::vector<WavefunctionFirstDerivativeType*> gradLogPsi;
+  std::vector<WavefunctionSecondDerivativeType*> lapLogPsi;
+
+  J1Spin(const std::string& obj_name, const ParticleSet& ions, ParticleSet& els)
+      : WaveFunctionComponent("J1Spin", obj_name),
+        myTableID(els.addTable(ions)),
+        Nions(ions.getTotalNum()),
+        Nelec(els.getTotalNum()),
+        NumGroups(determineNumGroups(ions)),
+        NumTargetGroups(determineNumGroups(els)),
+        Ions(ions),
+        Elecs(els),
+        NumVars(0)
+  {
+    if (myName.empty())
+      throw std::runtime_error("J1Spin object name cannot be empty!");
+    initialize();
+  }
+
+  J1Spin(const J1Spin& rhs) = delete;
+
+  ~J1Spin() override
+  {
+    delete_iter(gradLogPsi.begin(), gradLogPsi.end());
+    delete_iter(lapLogPsi.begin(), lapLogPsi.end());
+  }
+
+  /* determine NumGroups which controls the use of optimized code path using ion groups or not */
+  static int determineNumGroups(const ParticleSet& ions)
+  {
+    const int num_species = ions.getSpeciesSet().getTotalNum();
+    if (num_species == 1)
+      return 1;
+    else if (num_species > 1 && !ions.IsGrouped)
+      return 0;
+    else
+      return num_species;
+  }
+
+  /* initialize storage */
+  void initialize()
+  {
+    J1Functors.resize(Nions * Nelec, nullptr);
+    Vat.resize(Nelec);
+    Grad.resize(Nelec);
+    Lap.resize(Nelec);
+
+    U.resize(Nions);
+    dU.resize(Nions);
+    d2U.resize(Nions);
+    d3U.resize(Nions);
+    DistCompressed.resize(Nions);
+    DistIndice.resize(Nions);
+  }
+
+  void addFunc(int source_type, std::unique_ptr<FT> afunc, int target_type = -1)
+  {
+    // if target type is not specified J1Functors[i,:] is assigned
+    // if target type is specified J1Functors[i,j] is assigned
+    //
+    // if target type is not specified J1UniqueFunctors["i"] is assigned
+    // if target type is specified J1UniqueFunctors["ij"] is assigned
+    //
+    app_log() << " source" << source_type << " target type " << target_type << std::endl;
```

Comment:
If the electron species is not give, we can populate every target species with clones of the same unique_ptr.
I hope we only need J1UniqueFunctors and never need J1Functors.
I think `Unique` is with respect to particle groups. Spin symmetry generally should be handled in the builder not inside this function.

For example, if you have spin-up, down electrons and positrons, you would like to populate up/down electron with the same Jastrow but positrons with a different one, I don't see the symmetry can be well recorded in J1UniqueFunctors.
For this reason, my assessment on J1UniqueFunctors is that it is to represent groups.

Inside unique J1 functor, it should carry shared_ptr for optimization info. The symmetry can be linked there and builder should handle this.

Reply:
> I hope we only need J1UniqueFunctors and never need J1Functors. 
I think this might be possible. I need more time

Reply:
> Inside unique J1 functor, it should carry shared_ptr for optimization info. The symmetry can be linked there and builder should handle this.

Let me know if the current iteration of this is similar to what you are expecting.